test/test_tagging.py

```diff
@@ -0,0 +1,222 @@
+import json
+import unittest
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+from azul.deployment import (
+    populate_tags,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from azul_test_case import (
+    AzulUnitTestCase,
+)
+
+
+def setupModule():
+    configure_test_logging()
+
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def permissive_compare(self, expected, actual) -> bool:
+        """
+        Wrapper for assertDictEqualPermissive doctests.
+
+        >>> t = TestTerraformResourceTags()
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]}
+        ... )
+        True
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        True
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        False
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]}
+        ... )
+        False
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': None, 456: 'def'},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        False
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': None},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        True
+        """
+        try:
+            self.assertDictEqualPermissive(expected, actual)
+        except AssertionError:
```

Comment:
This is weird. 

I was under the impression that we agreed that the helper method returns bool and contains the meat of the functionality as well as the doctests. Did we misunderstand each other or did you intentionally deviate from the agreement for a particular reason? If the latter, you should pro-actively draw my attention to that and outline your reasoning.

Reply:
I think we misunderstood each other. I neglected to record the exact details of our conversation in my notes, but from what I remember:

* We agreed that `self.assert*` should be used to make it easier to identify deviations between the expected and observed output.
* We also agreed that writing out `AssertionError`s in doctests would be verbose.

My implementation was largely guided by these two points.

Reply:
Yes, those two bulleted items were part of our agreement. I also believe that we agreed to write a worker method that returns a boolean and a wrapper method that calls that worker method and raises an assertion error if the return value is False. The current implementation does the inverse, so to speak. The "pretty diff" directive gets in the way of the "boolean worker" directive which is probably why you went that route.

Let's tweak this a little bit. Instead of True/False, have the worker return the path of the discrepancy or None if there is no significant difference. Include the path in the assertion error that the wrapper raises. I hope that's going to simplify things.